File: Go-Solutions/16/16.go
package main

import (
	"math"
	"sort"
)

func main() {
	
}

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	ans := nums[0] + nums[1] + nums[2]
	for i := 0; i < len(nums) - 2; i++ {
		start := i + 1
		end := len(nums) - 1
		for start < end {
			localSum := nums[i] + nums[start] + nums[end]
			if int(math.Abs(float64(localSum - target))) < int(math.Abs(float64(ans - target))) {
				ans = localSum
			}
			if localSum < target {
				start++
			} else if localSum > target {
				end--
			} else {
				return ans
			}
		}
	}
	return ans
}

File: Java-Kotlin/src/main/java/array/_16/_16.java
package array._16;

import java.util.Arrays;

public class _16 {

    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int ans = nums[0] + nums[1] + nums[2];
        int len = nums.length;
        for (int i = 0; i < len - 2; i++) {
            int start = i + 1;
            int end = len - 1;
            while (start < end) {
               int localSum = nums[i] + nums[start] + nums[end];
               if (Math.abs(localSum - target) < Math.abs(ans - target)) {
                   ans = localSum;
               }
               if (localSum > target) {
                   end--;
               } else if (localSum < target) {
                   start++;
               } else {
                   return ans;
               }
            }
        }
        return ans;
    }
}
